### PR TITLE
Dup default property values

### PR DESCRIPTION
**Pull request by lamont-granquist**
This largely makes frozen default value errors go away.  Each
copy of a resource will get a dup of the default.  Defaults are
still frozen so we have defense-in-depth against them being
mutated.  This eliminates the need to `default: lazy{ {} }`
things like Hashes or Arrays.

### CODE REVIEW

File: spec/unit/property_spec.rb [hunk 1]
```diff
@@ -708,10 +709,10 @@ def blah
           expect(resource.x).to eq "hi1"
           expect(Namer.current_index).to eq 1
         end
-        it "when x is retrieved, coercion is run each time" do
+        it "when x is retrieved, coercion is run exactly once" do
           expect(resource.x).to eq "101"
-          expect(resource.x).to eq "102"
-          expect(Namer.current_index).to eq 2
+          expect(resource.x).to eq "101"
+          expect(Namer.current_index).to eq 1
```
Comment (lamont-granquist) [inline spec/unit/property_spec.rb:715]: This is one of the weirdest tests I've ever seen in the codebase and I'm unsure of what its purpose is, or if it was just sort of documentation-through-code of an edge condition.

**Comment by tas50** [comment]
Add a link to the 17 section of the release notes to this PR so I can put them together come april

File: lib/chef/property.rb [hunk 2]
```diff
@@ -748,5 +749,22 @@ def coerce_and_validate(resource, value)
 
       result
     end
+
+    # recursively dup the value
+    def deep_dup(value)
```
Comment (jaymzh) [inline lib/chef/property.rb:754]: Do we really not have a deep_dup somewhere? We do this in the Node constnatly, right?

File: spec/unit/property_spec.rb (hunk 1 above)
Comment (jaymzh) [reply spec/unit/property_spec.rb:715]: we are changing the behavior here.... which this behavior seems better.... but like this will break SOMETHING. We'll need to document that.

Comment (lamont-granquist) [reply spec/unit/property_spec.rb:715]: I literally have no idea what this would ever break though or why anyone would ever have done this.

File: lib/chef/property.rb (hunk 2 above)
Comment (lamont-granquist) [reply lib/chef/property.rb:754]: That is all overly complicated stuff hanging off of the various attribute classes, which would need to be injected into Hash and Array to function similarly.

**Review by jaymzh — APPROVED** [review]
